### ISSUE TITLE
Is it possible to add an entry to a table if it has 0 key fields?

### ISSUE THREAD

**Issue by jafingerhut** [p4lang/p4runtime#87]
This question is relevant to _any_ control plane API for configuring the tables in a P4 program, but if it is easier to answer in the context of the P4Runtime API specifically, that is enough for now.

The reason why I am bothering to ask: Working on a P4 test packet generation tool, and want to ensure that I am not missing any valid test cases, nor generating extra ones unnecessarily.

The open source P4 compilers often create tables that do not exist in the original P4 program that I will call 'dummy tables' here, although there is no official name for these that I know of in the language specification.  These have have no fields in their keys.

My understand is that the P4Runtime API should reject as an error any message that one might attempt to create that could add an entry to such a table, either because:

+ the message mentions some key fields, which are an error because the table has none

or

+ the message mentions no key fields, and thus if it is a 'table insert' should fail because only 'table modify' is supported by P4Runtime API for a table's default actions.

Does all of that sound correct?

**Comment by antoninbas** [comment]
This is correct.
Note that the P4Runtime specification currently does not specify the behavior of the server is the client attempts to add entries past the table capacity. I bring this up because as far as I'm concerned the only valid size value for a keyless table is 0, which means that someone attempting to add an entry to a keyless table should probably just get the same error code as when trying to add entries after the table is full.
The issue is that we haven't defined what constitutes a table's maximum capacity. Should we return a "table full" error if we reach the size specified in the P4 program (like bmv2 does) or should we try to add the entry until we run out of HW resources for that table (the allocated HW resource may be > than the P4 size).

**Comment by jafingerhut** [comment]
I see how questions of maximum table capacity arise from my question, but I was hoping not to go there today :-)

I agree the only reasonable size for a keyless table is 0, too.  A warning or error from the compiler front end seems reasonable if a keyless table tries to specify a size other than 0.

Even if we don't get into maximum table sizes, as long as P4Runtime servers will give _some_ kind of error when attempting to add a regular table entry to a keyless table, I am content.

If we want a separate issue on the question of maximum table sizes, I'd propose a separate issue for that, but don't have many current thoughts on the topic that might push it forward.

If we are all agreed on the original intent of this question, I am happy to close this issue.  Thanks for the answer.

**Comment by antoninbas** [comment]
Yes, that was probably uncool of me to bring it up in this issue :) Actually I do believe that with the current state of things for P4 & PSA, the bmv2 behavior makes sense (RESOURCE_EXHAUSTED error if the client tries to go above the limit). AFAIK there has been some discussions about adding complexity and usefulness to size specification (min guaranteed size, and maybe more), and if these discussions do go through, it would definitely make sense to think about refining the P4Runtime spec. As things stand today, I suggest going with the bmv2 behavior for P4Runtime 1.0.

**Comment by jafingerhut** [comment]
No worries on any possible uncoolness.

I am one of the people pushing for the `guaranteed_size = true` attribute.  Not hard, but gently and persistently.

So far I can't personally think of much reason to want to enforce maximum number of entries in a table.  If the device can do it, I say let the controller do it!  But I would also not be surprised if someone had a use case that would convince me, if explained in a way that I got it.

Closing this issue, since my original question was answered.
